src/ui/groups_widgets.cc

```diff
@@ -790,34 +797,82 @@ void DictGroupsWidget::addAutoGroupsByFolders()
     return;
   }
 
-  QMultiMap< QString, sptr< Dictionary::Class > > dictMap;
+  // Map from dict to ContainerFolder's parent
+  QMap< sptr< Dictionary::Class >, QString > dictToContainerFolder;
 
   for ( const auto & dict : *activeDicts ) {
-    QString dictFolder = dict->getContainingFolder();
+    QDir c = dict->getContainingFolder();
 
-    if ( dictFolder.isEmpty() ) {
-      continue;
+    if ( !c.cdUp() ) {
+      cdUpWentWrong( c.absolutePath() );
+      return;
     }
 
-    QDir dir = dictFolder;
-    if ( dir.cdUp() ) {
-      dictMap.insert( dir.dirName(), dict );
+    dictToContainerFolder.insert( dict, c.absolutePath() );
+  }
+
+  /*
+  Check for duplicated ContainerFolder's parent folder names, and prepend the upper parent's name.
+
+  .
+  ├─epistularum
+  │   └─Japanese   <- Group
+  │       └─DictA  <- Dict Files's container folder
+  |          └─ DictA Files
+  ├─Mastameta
+  │   └─Japanese   <- Group
+  |       └─DictB  <- Dict Files's container folder
+  |          └─ DictB Files
+
+  will be grouped into epistularum/Japanese and Mastameta/Japanese
+  */
+
+  // set a dirname to true if there are duplications like `epistularum/Japanese` and `Mastameta/Japanese`
+  QHash< QString, bool > dirNeedPrepend{};
+
+  for ( const auto & path : dictToContainerFolder.values() ) {
+    auto dir = QDir( path );
+    if ( dirNeedPrepend.contains( dir.dirName() ) ) {
+      dirNeedPrepend[ dir.dirName() ] = true;
     }
     else {
-      qWarning() << "Cannot auto group because parent folder cannot be reached: " << dir;
-      continue;
+      dirNeedPrepend.insert( dir.dirName(), false );
+    }
+  }
+
+  // map from GroupName to dicts
+  QMultiMap< QString, sptr< Dictionary::Class > > groupToDicts;
+
+  for ( auto & dict : dictToContainerFolder.keys() ) {
+    QDir path = dictToContainerFolder[ dict ];
+    QString groupName;
+    if ( !dirNeedPrepend[ path.dirName() ] ) {
+      groupName = path.dirName();
     }
+    else {
+      QString directFolder = path.dirName();
+      if ( !path.cdUp() ) {
+        cdUpWentWrong( path.absolutePath() );
+        return;
+      }
+      QString upperFolder = path.dirName();
+      groupName           = upperFolder + "/" + directFolder;
+    }
+
+    groupToDicts.insert( groupName, dict );
   }
 
-  for ( const auto & group : dictMap.uniqueKeys() ) {
+  // create and insert groups
+  // modifying user's groups begins here
+  for ( const auto & group : groupToDicts.uniqueKeys() ) {
     if ( count() != 0 ) {
       setCurrentIndex( count() - 1 );
     }
 
     addUniqueGroup( group );
     DictListModel * model = getCurrentModel();
 
```

Comment:
`group ` may already existed.

Reply:
The `addUniqeGroup` is misleading. It will also check if an existing one exist, if it does, then set the current index to there.

So it doesn't matter if already exist.

I copied code from "auto group by languages"

https://github.com/xiaoyifang/goldendict-ng/blob/f4e230b2810b4f520d3ea8e32ef6bb48fc6aa655/src/ui/groups_widgets.cc#L759-L773